tracer/src/Datadog.Trace/Sampling/RuleBasedSampler.cs

```diff
@@ -77,21 +76,23 @@ public void RegisterRule(ISamplingRule rule)
             _rules.Add(rule);
         }
 
-        private int GetSamplingPriority(Span span, float rate, bool agentSampling)
+        private SamplingDecision MakeSamplingDecision(Span span, float rate, int mechanism)
         {
             // make a sampling decision as a function of traceId and sampling rate
             var sample = ((span.TraceId * KnuthFactor) % TracerConstants.MaxTraceId) <= (rate * TracerConstants.MaxTraceId);
 
-            // legacy sampling based on data from agent
-            if (agentSampling)
-            {
-                return sample ? SamplingPriorityValues.AutoKeep : SamplingPriorityValues.AutoReject;
-            }
+            var priority = mechanism switch
+                           {
+                               // default sampling rule based on sampling rates from agent response.
+                               // if sampling decision was made automatically without any input from user, use AutoKeep/AutoReject.
+                               SamplingMechanism.AgentRate => sample ? SamplingPriorityValues.AutoKeep : SamplingPriorityValues.AutoReject,
+
+                               // sampling rule based on user configuration (DD_TRACE_SAMPLE_RATE, DD_TRACE_SAMPLING_RULES).
+                               // if user influenced sampling decision in any way (manually, rules, rates, etc), use UserKeep/UserReject.
+                               _ => sample && _limiter.Allowed(span) ? SamplingPriorityValues.UserKeep : SamplingPriorityValues.UserReject
```

Comment:
What about the other sampling mechanisms, like `Asm`?

Reply:
Not all sampling decisions use this sampler. ASM for example just sets `UserKeep` (now with sampling mechanism `Asm`) on traces they decide to keep.